src/main/java/in/ekstep/am/step/TokenSignStep.java

```diff
@@ -1,190 +1,249 @@
-            long refreshTokenLogOlderThan = Long.parseLong(keyManager.getValueFromKeyMetaData("refresh.token.log.older.than"));
-            long diffInDays = (currentTime - tokenWasIssuedAt) / (60 * 60 * 24);
-            if (diffInDays >= refreshTokenLogOlderThan) {
-                log.info("Token issued before: " + diffInDays + ", UID: " + body.get("sub") + ", aud: " + body.get("aud") + ", exp: " + body.get("exp") + ", iat: " + body.get("iat"));
-            }
-            else {
-                log.info("Token issued for UID: " + body.get("sub") + ", aud: " + body.get("aud") + ", exp: " + body.get("exp") + ", iat: " + body.get("iat"));
-            }
-        }
-        else {
-            log.info("Token issued for UID: " + body.get("sub") + ", aud: " + body.get("aud") + ", exp: " + body.get("exp") + ", iat: " + body.get("iat"));
-        }
+    if(bodyData.get("typ").equals("Offline")) {
+      tokenSignResponseBuilder.setRefreshExpiresIn(0);
+    }
+    else {
+      tokenSignResponseBuilder.setRefreshExpiresIn(tokenExpiry - currentTime);
     }
 
-    @Override
-    public void execute() {
-        if(isJwtTokenValid())
-            generateNewJwtToken();
-        else
-            tokenSignResponseBuilder.markFailure("invalid_grant", "invalid_grant");
+    if(bodyData.get("typ").equals("Offline")) {
+      long refreshTokenLogOlderThan = Long.parseLong(keyManager.getValueFromKeyMetaData("refresh.token.log.older.than"));
+      long diffInDays = (currentTime - tokenWasIssuedAt) / (60 * 60 * 24);
+      if (diffInDays >= refreshTokenLogOlderThan) {
+        log.info("Token issued before: " + diffInDays + ", UID: " + body.get("sub") + ", aud: " + body.get("aud") + ", exp: " + body.get("exp") + ", iat: " + body.get("iat"));
+      }
+      else {
+        log.info("Token issued for UID: " + body.get("sub") + ", aud: " + body.get("aud") + ", exp: " + body.get("exp") + ", iat: " + body.get("iat"));
+      }
+    }
+    else {
+      log.info("Token issued for UID: " + body.get("sub") + ", aud: " + body.get("aud") + ", exp: " + body.get("exp") + ", iat: " + body.get("iat"));
+    }
+  }
+
+  private List<Map<String,Object>> appendRoles(AmResponse response) {
+    if (200 == response.code()) {
+      Map<String,Object> learnerResponse = GsonUtil.fromJson(response.body(),Map.class);
+      if (MapUtils.isNotEmpty(learnerResponse)) {
+        Map<String,Object> result = (Map<String, Object>) learnerResponse.get("result");
+        if (MapUtils.isNotEmpty(result)) {
+          return  (List<Map<String,Object>>)result.get("roles");
+        }
+      }
+    } else {
+      log.info("Got error response from learner api : "+response.body());
+    }
+    return appendDefaultRoles();
+  }
+
+  private List<Map<String, Object>> appendDefaultRoles() {
+    //if user role empty , pass PUBLIC Role
+    List<Map<String,Object>> roles = new ArrayList<>();
+    Map<String,Object> role = new HashMap<>();
+    role.put("role","PUBLIC");
```

Comment:
- Default roles are not being populated in case the user role list is empty - Example I try to request refresh token with a keycloak service account token, where the userid will not be found in cassandra.
- In case of a 404 also default roles are not being appended
This is something we should discuss and see if the use case requires default roles